test/compress/collapse_vars.js

```diff
@@ -68,11 +68,10 @@ collapse_vars_side_effects_1: {
             log(x, s.charAt(i++), y, 7);
         }
         function f4() {
-            var log = console.log.bind(console),
-                i = 10,
+            var i = 10,
                 x = i += 2,
                 y = i += 3;
-            log(x, i += 4, y, i);
+            console.log.bind(console)(x, i += 4, y, i);
```

Comment:
It does this:
- begin scanning on `log`
- collect all assignments within the value (none in this case, but if say `log = i = 10` it would have marked `i` down)
- realise `console.log.bind(console)` may have side-effects
- every time it hits `AST_SymbolRef` (`i` in this case), make sure all references to `i` are within this scope, i.e. `f4()`

So `collapse_vars` move past each `i` safely because it knows whatever side-effects `console.log.bin(console)` may have, it can't possibly modify `i` which is exclusively modified within `f4()`.